python/vast/utils/arrow.py

```diff
@@ -1,7 +1,72 @@
-import ipaddress
+import ipaddress as ip
+from typing import SupportsBytes
 
 import pyarrow as pa
 
+# Accepts a 128-bit buffer holding an IPv6 address and returns an IPv4 or IPv6
+# address.
+def unpack_ip(buffer: SupportsBytes) -> ip.IPv4Address | ip.IPv6Address:
+    num = int.from_bytes(buffer, byteorder="big")
+    # Convert IPv4 mapped addresses back to regular IPv4.
+    # https://tools.ietf.org/html/rfc4291#section-2.5.5.2
+    if (num >> 32) == 65535:
+        num = num - (65535 << 32)
+    return ip.ip_address(num)
+
+
+class IPAddressScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
+        return unpack_ip(self.value.as_py())
+
+
+class IPAddressType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.binary(16), "vast.address")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized):
+        return IPAddressType()
```

Comment:
These should match their equivalents in libvast:

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/src/type.cpp#L1913-L1926